business_rules/engine.py

```diff
@@ -78,21 +141,42 @@ def _do_operator_comparison(operator_type, operator_name, comparison_value):
     comparison_value is whatever python type to compare to
     returns a bool
     """
+
     def fallback(*args, **kwargs):
         raise AssertionError("Operator {0} does not exist for type {1}".format(
             operator_name, operator_type.__class__.__name__))
+
     method = getattr(operator_type, operator_name, fallback)
     if getattr(method, 'input_type', '') == FIELD_NO_INPUT:
         return method()
     return method(comparison_value)
 
 
-async def do_actions(actions, defined_actions):
+def do_actions(actions, defined_actions) -> dict:
+    """ Run the actions
+    Attributes:
+        actions(list): list of dictionaries of actions. e.g: [
+            { "name": "put_on_sale",
+            "params": {"sale_percentage": 0.25},
+            }
+        ]
+    Returns:
+        actions_results(dict): Dictionary of actions results
+            e.g: {"put_on_sale: [product1, product2, ...]}
+    """
+    actions_results = {}
     for action in actions:
         method_name = action['name']
-        def fallback(*args, **kwargs):
-            raise AssertionError("Action {0} is not defined in class {1}"\
-                    .format(method_name, defined_actions.__class__.__name__))
+
         params = action.get('params') or {}
-        method = getattr(defined_actions, method_name, fallback)
-        await method(**params)
+        if hasattr(defined_actions, method_name):
+            method = getattr(defined_actions, method_name)
+        else:
+            raise AssertionError(
+                'Action {} is not defined in class {}'.format(
+                    method_name, defined_actions.__class__.__name__
+                )
+            )
+        actions_results[method_name] = method(**params)
```

Comment:
there can be several the same actions, but with different params.
As it stored in dictionary with `method_name` as a key, it will be overwritten.
better use list of dicts with method name and method params.

Reply:
Not in the same class!